lib/package.js

```diff
@@ -140,66 +140,67 @@ exports.lookup = function(pkg) {
 };
 
 exports.getVersionMatch = getVersionMatch;
-function getVersionMatch(pkgVersion, versions) {
+function getVersionMatch(pkgVersion, versions, options) {
   // unescape pkgVersion for comparison
   if (pkgVersion)
     pkgVersion = decodeURIComponent(pkgVersion);
 
-  var latestPre;
-  var versionList = [];
-  var exactVersions = [];
+  var version;
+  var stableVersions = [];
+  var unstableVersions = [];
 
   Object.keys(versions).forEach(function(v) {
-    versions[v].version = v;
-    if (versions[v].stable === false || versions[v].exactOnly)
-      exactVersions.push(v);
-    else
-      versionList.push(v);
-  });
-
-  exactVersions.sort(semver.compare).reverse();
-  versionList.sort(semver.compare).reverse();
-
-  // find highest stable match in tags
-  for (var i = 0; i < versionList.length; i++) {
-    var version = versionList[i];
-    var semverMatch = version.match(semver.semverRegEx);
+    version = versions[v];
+    var stable = version.stable;
+    var semverMatch = v.match(semver.semverRegEx);
     var valid = semverMatch && semverMatch[1] && semverMatch[2] && semverMatch[3];
     var pre = valid && semverMatch[4];
-    var stable = versions[version].stable;
+
+    // store a reverse lookup
+    version.version = v;
 
     // ignore non-semver or prerelease, unless explictly marked as stable
-    if (!stable && (!valid || pre)) {
-      latestPre = latestPre || pre && version;
-      continue;
-    }
+    if (stable === false || !valid || (stable !== true && pre))
+      unstableVersions.push(v);
+    else
+      stableVersions.push(v);
+  });
 
-    if (!pkgVersion || semver.match(pkgVersion, version))
-      return versions[version];
+  function compareDesc(a, b) {
+    return semver.compare(b, a);
   }
 
-  // if we asked for latest, and nothing stable found, use master or otherwise top
   if (!pkgVersion) {
-    if (latestPre)
-      return versions[latestPre];
-    if (versions.master)
-      return versions.master;
-    return versions[versionList[0] || exactVersions[0]];
-  }
-
-  // next try an unstable version range match in tags then branches
-  for (var j = 0; j < versionList.length; j++) {
-    if (semver.match(pkgVersion, versionList[j]))
-      return versions[versionList[j]];
+    var latest = options && options.latestVersion && versions[options.latestVersion];
+    if (latest && latest.stable !== false) return latest;
```

Comment:
Perhaps we should take into account default stability as well here - `1.0.0-beta` is not stable, even if latest, unless explicitly overriden?